src/connectivity/components.c

```diff
@@ -384,39 +390,48 @@ static igraph_error_t igraph_is_connected_weak(const igraph_t *graph, igraph_boo
 igraph_error_t igraph_is_connected(const igraph_t *graph, igraph_bool_t *res,
                         igraph_connectedness_t mode) {
 
+    igraph_cached_property_t prop;
     igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_error_t retval = IGRAPH_SUCCESS;
+    igraph_integer_t no;
+
+    switch (mode) {
+        case IGRAPH_WEAK:
+            prop = IGRAPH_PROP_IS_WEAKLY_CONNECTED;
+            break;
+
+        case IGRAPH_STRONG:
+            prop = IGRAPH_PROP_IS_STRONGLY_CONNECTED;
+            break;
+
+        default:
+            IGRAPH_ERROR("Invalid connectedness mode.", IGRAPH_EINVAL);
+    }
+
+    IGRAPH_RETURN_IF_CACHED_BOOL(graph, prop, res);
```

Comment:
If the graph is undirected, it should always consider `IGRAPH_PROP_IS_WEAKLY_CONNECTED`, no?

Why not do the caching directly in `igraph_is_connected_weak` and `igraph_i_connected_components_strong`?

Reply:
Good point about the undirected case, I've fixed that in 3189a7229

As for the second question, I wanted to have a top-level function that deals with the caching and then delegates the grunt work to lower-level functions that then do _not_ need to be concerned with caching.